text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,365 @@
+could implement `Copy + DerefMut` and actually rely on the deref coercion inside
+a union.  That seems very unlikely, but should be tested with a crater run.
+
+The initialization tracking rules are somewhat surprising, and one might prefer
+the compiler to just not track anything when it comes to unions.  After all, the
+compiler fundamentally cannot know what part of the union is properly
+initialized.  Unfortunately, not having any initialization tracking is not an
+option when non-`Copy` fields are involved: We have to decide if moving out of a
+union field is allowed.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Ruling out fields with drop glue does not, in fact, reduce the expressiveness of
+unions because one can use `ManuallyDrop<T>` to obtain a drop-glue-free version
+of `T`.  If anything, having the `ManuallyDrop` in the union definition should
+help to drive home the point that no automatic dropping is happening, ever.
+(Before this RFC, automatic dropping is happening when assigning to a union
+field but not when the union goes out of scope.  That seems to be the result of
+necessity, not of a coherent design.)
+
+An alternative approach to proceed with unions has been
+[previously proposed by @petrochenkov](https://github.com/petrochenkov/rfcs/blob/e5266bd105f592f7408b8592c5c3deaccba7f1ec/text/1444-union.md#initialization-state).
+That proposal replaces RFC 1444 and goes into a lot more points than this much more
+limited proposal.  In particular, it allows fields with drop glue.  However, it
+can be pretty hard for the programmer to predict when drop glue will be
+automatically invoked on assignment or not, because the initialization tracking
+(which this RFC adapts from @petrochenkov's proposal) can sometimes be a little
+surprising when looking at individual fields: Whether `u.f2 = ...;` drops
+depends on whether `u.f1` has been previously initialized.  @petrochenkov hence
+proposes a lint to warn people that unions with drop-glue fields are not always
+very well-behaved.  This RFC, on the other hand, side-steps the entire question
+by not allowing fields with drop glue.  Initialization tracking thus has no
+effect on the code executed during an assignment of a union field.  For unions
+that `impl Drop`, it still has an effect on what happens when the union goes out
+of scope, but in that case initialization is so restricted that I cannot think
+of any surprises.  Together with the `DerefMut` restriction, that should make it
+very unlikely to accidentally call `drop` when it was not intended.
+
+We could significantly simplify the initialization tracking by always applying
+the rules that are currently only applied to unions that `impl Drop`.  However,
+that does not actually help with the pitfall described above.  The more complex
+rules allow more code that many will reasonably expect to work, and do not seem
+to introduce any additional pitfalls.
```

Comment:
Independent nested fields borrowed independently is probably the most natural expectation:
```rust
union U { a: (u8, u8), b: u16 }

let x = &u.a.0;
let y = &u.a.1;
```
so it's *borrow* checker that needs to work in per-field fashion first of all.
Move checker can just mirror what borrow checker does (for consistency and also because they share common infrastructure in the compiler).

Reply:
Ah, I hadn't thought about borrowck here. So with the rules as stated, once a field is (partially) borrowed, its siblings all become completely blocked from borrowing?

Borrowing is unsafe, so this would not be necessary, but it still seems useful.